### PR TITLE
couch sql migrate

### PR DESCRIPTION
**Pull request by snopoke**
@czue

buddy @biyeun

Sorry this got a bit big:
- [ ] print migration stats
- [ ] save document diffs and allow printing them
- [ ] do timezone migration
- [ ] prevent form submissions during migration
- [ ] commit migration once complete
- [ ] migrate attachments & form operations
- [ ] copy unprocessed forms

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/couch_sql_migration/couchsqlmigration.py, line 78 - E501 line too long (121 > 115 characters)
* corehq/apps/couch_sql_migration/couchsqlmigration.py, line 377 - W391 blank line at end of file
* corehq/apps/couch_sql_migration/management/commands/migrate_domain_from_couch_to_sql.py, line 77 - E501 line too long (119 > 115 characters)
* corehq/apps/couch_sql_migration/management/commands/migrate_domain_from_couch_to_sql.py, line 85 - E501 line too long (118 > 115 characters)
* corehq/apps/tzmigration/timezonemigration.py, line 18 - F401 'iso_string_to_datetime' imported but unused

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/couch_sql_migration/couchsqlmigration.py, line 79 - E501 line too long (121 > 115 characters)
* corehq/apps/couch_sql_migration/couchsqlmigration.py, line 378 - W391 blank line at end of file
* corehq/apps/couch_sql_migration/management/commands/migrate_domain_from_couch_to_sql.py, line 77 - E501 line too long (119 > 115 characters)
* corehq/apps/couch_sql_migration/management/commands/migrate_domain_from_couch_to_sql.py, line 85 - E501 line too long (118 > 115 characters)
* corehq/apps/tzmigration/timezonemigration.py, line 18 - F401 'iso_string_to_datetime' imported but unused

**Comment by czue** [comment]
wowzers this looks monstrous. haven't looked closely yet, but build is failing. should i review :whale2: ?

```
======================================================================
ERROR: test_migration (corehq.apps.tzmigration.tests.test_timezonemigration.TimeZoneMigrationTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/tzmigration/tests/test_timezonemigration.py", line 102, in test_migration
    run_timezone_migration_for_domain(self.domain)
  File "/mnt/commcare-hq-ro/corehq/apps/tzmigration/timezonemigration.py", line 26, in run_timezone_migration_for_domain
    _run_timezone_migration_for_domain(domain)
  File "/mnt/commcare-hq-ro/corehq/apps/tzmigration/timezonemigration.py", line 81, in _run_timezone_migration_for_domain
    planning_db = prepare_planning_db(domain)
  File "/mnt/commcare-hq-ro/corehq/apps/tzmigration/timezonemigration.py", line 152, in prepare_planning_db
    planning_db.add_diffs('form', xform_id,
AttributeError: 'PlanningDB' object has no attribute 'add_diffs'
----------------------------------------------------------------------
```
```
======================================================================
FAIL: test_basic_case_migration (corehq.form_processor.tests.test_migration.MigrationTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/form_processor/tests/test_migration.py", line 28, in test_basic_case_migration
    self._do_migration_and_assert_flags(self.domain)
  File "/mnt/commcare-hq-ro/corehq/form_processor/tests/test_migration.py", line 35, in _do_migration_and_assert_flags
    self.assertTrue(should_use_sql_backend(domain))
AssertionError: None is not true
-------------------- >> begin captured logging << --------------------
```

**Comment by snopoke** [comment]
Fine to review. I'll fix the build.

**Comment by czue** [comment]
sorry question was commit or at once

**Comment by snopoke** [comment]
Hmm, lot's of commits i know but probably easier per commit.

File: corehq/apps/couch_sql_migration/management/commands/migrate_domain_from_couch_to_sql.py [hunk 1]
```diff
@@ -65,3 +73,30 @@ def _print_status(self, name, ids_in_couch, ids_in_sql):
...
+    if raw_input(
+            '{} [y/n]'.format(message)
+    ).lower() == 'y':
+        return
+    else:
+        raise CommandError('abort')
+
+
+def _blow_away_migration(domain):
+    assert not should_use_sql_backend(domain)
```
Comment (czue) [inline corehq/apps/couch_sql_migration/management/commands/migrate_domain_from_couch_to_sql.py]: :+1:

File: corehq/apps/couch_sql_migration/couchsqlmigration.py [hunk 2]
```diff
@@ -264,3 +265,28 @@ def delete_diff_db(domain):
         # reraise on any other error
         if e.errno != 2:
             raise
+
+
+BASE_IGNORED_FORM_PATHS = {
+    '_rev',
+    'migrating_blobs_from_couch',
+    '#export_tag',
+    'computed_',
```
Comment (czue) [inline corehq/apps/couch_sql_migration/couchsqlmigration.py]: haha forgot about this thing. we should :skull: sometime (once MVP reports are also gone)

File: corehq/apps/couch_sql_migration/couchsqlmigration.py [hunk 3]
```diff
@@ -264,3 +265,28 @@ def delete_diff_db(domain):
...
+    'state',
+    'edited_on',
+    'computed_modified_on_',
+    'problem',
+    'orig_id',
+    'deprecated_form_id',
```
Comment (czue) [inline corehq/apps/couch_sql_migration/couchsqlmigration.py]: for some of these we should still find a way to compare the couch and sql versions if we're not already

Comment (snopoke) [inline corehq/apps/couch_sql_migration/couchsqlmigration.py]: Yes, some of them like 'orig_id' and 'problem' should be compared only for
specific doc types

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/couch_sql_migration/couchsqlmigration.py, line 79 - E501 line too long (121 > 115 characters)
* corehq/apps/couch_sql_migration/couchsqlmigration.py, line 378 - W391 blank line at end of file
* corehq/apps/couch_sql_migration/management/commands/migrate_domain_from_couch_to_sql.py, line 77 - E501 line too long (119 > 115 characters)
* corehq/apps/couch_sql_migration/management/commands/migrate_domain_from_couch_to_sql.py, line 85 - E501 line too long (118 > 115 characters)
* corehq/apps/tzmigration/timezonemigration.py, line 18 - F401 'iso_string_to_datetime' imported but unused

**Comment by czue** [comment]
wow, this is looking good - well done

**Comment by snopoke** [comment]
Well done for getting through the review!

**Comment by benrudolph** [comment]
oh man! it's happening

**Pull request merged**